src/transport/SessionManager.cpp

```diff
@@ -573,20 +579,106 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
         }
         else
         {
-            SecureUnicastMessageDispatch(partialPacketHeader, peerAddress, std::move(msg));
+            SecureUnicastMessageDispatch(partialPacketHeader, peerAddress, std::move(msg), ctxt);
         }
     }
     else
     {
-        UnauthenticatedMessageDispatch(partialPacketHeader, peerAddress, std::move(msg));
+        UnauthenticatedMessageDispatch(partialPacketHeader, peerAddress, std::move(msg), ctxt);
+    }
+}
+
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void SessionManager::SetConnectionCallbacks(OnConnectionCompleteCallback connCompleteCb, OnConnectionClosedCallback connClosedCb,
+                                            void * connContext)
+{
+    mConnCompleteCb = connCompleteCb;
+    mConnClosedCb   = connClosedCb;
+    mConnContext    = connContext;
+}
+
+void SessionManager::HandleConnectionComplete(Transport::ActiveConnectionState * conObj, CHIP_ERROR conErr)
+{
+    if (mConnCompleteCb != nullptr)
+    {
+        ChipLogProgress(Inet, "Calling Connection Complete callback");
+        mConnCompleteCb(mConnContext, conObj, conErr);
+    }
+    else
+    {
+        ChipLogProgress(Inet, "Connection Complete callback missing");
+    }
+}
+
+void SessionManager::HandleConnectionClosed(Transport::ActiveConnectionState * conObj, CHIP_ERROR conErr)
+{
+    if (mConnClosedCb != nullptr)
+    {
+        mConnClosedCb(mConnContext, conObj, conErr);
+    }
+    else
+    {
+        ChipLogProgress(Inet, "Connection Closed callback missing");
+    }
+
+    // Mark the corresponding secure session as defunct
+    mSecureSessions.ForEachSession([&](auto session) {
+        if (session->IsActiveSession() && session->GetTCPConnection() == conObj)
+        {
+            // Dis-associate the connection from session by setting it to
+            // nullptr.
+            session->SetTCPConnection(nullptr);
+            // Mark session as defunct
+            session->MarkAsDefunct();
+            return Loop::Break;
```

Comment:
There should ideally be one active session over a given connection.

Reply:
"Ideally", or guaranteed?

Reply:
I would think it to be guaranteed. However, the number of sessions should be relatively small and it may not harm in continuing through the loop, just in case.